### PR TITLE
AP_RangeFinder: add driver for BlueRobotics Ping sensor

### PR DESCRIPTION
**Pull request by rmackay9**
This PR adds basic support for the [BlueRobotics Ping sensor](https://www.bluerobotics.com/store/sensors-sonars-cameras/sonar/ping-sonar-r2-rp/).

The [protocol for the sensor can be found here](https://github.com/bluerobotics/ping-protocol).

This has been bench tested (it produces odd values because the sensor is not in the water) but seems to function.

Some features we may want to add in follow-up PRs include:

- more complete initialisation of the sensor (this PR uses all defaults)
- default the baud rate to 115200 (currently it always uses the SERIALx_BAUD parameter)

Feedback from @jaxxzer or some of the other devs at BlueRobotics would be greatly appreciated!

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+        if (b < 0) {
+            continue;
+        }
+        if (parse_byte(b)) {
+            count++;
+            sum_cm += distance_cm;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: How about using a return parameter here rather than storing `distance_cm` in `*this`?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        break;
+
+    case MSG_STATE_LEN_L:
+        msg.payload_len = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_LEN_H;
+        break;
+
+    case MSG_STATE_LEN_H:
+        msg.payload_len += ((uint16_t)b << 8);
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: `|=`?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        break;
+
+    case MSG_STATE_MSG_ID_L:
+        msg.msgid = b;
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_MSG_ID_H;
+        break;
+
+    case MSG_STATE_MSG_ID_H:
+        msg.msgid += ((uint16_t)b << 8);
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: `|+`?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        }
+        break;
+
+    case MSG_STATE_CRC_L:
+        msg.crc = b;
+        msg.state = MSG_STATE_CRC_H;
+        break;
+
+    case MSG_STATE_CRC_H:
+        msg.crc += ((uint16_t)b << 8);
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: `|=`?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        msg.state = MSG_STATE_HEADER1;
+        if (msg.crc_expected == msg.crc) {
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: No diagnostics about parse errors in this driver.  *shrug*

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+            // process payload
+            switch(msg.msgid) {
+
+            case BLPING_MSGID_ACK:
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: This switch is a fair indication that this should be an `enum class`, not a list of defines.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+            case BLPING_MSGID_NACK:
+                // ignore
+                break;
+
+            case BLDPIN_MSGID_DISTANCE_SIMPLE:
+                uint32_t dist_mm = (uint32_t)msg.payload[0] | (uint32_t)msg.payload[1] << 8 | (uint32_t)msg.payload[2] << 16 | (uint32_t)msg.payload[3] << 24;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: `const`

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_BLPing.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <ctype.h>
+#include <AP_HAL/utility/sparse-endian.h>
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: Did you end up using anything from this header?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+#include "RangeFinder_Backend.h"
+
+class AP_RangeFinder_BLPing : public AP_RangeFinder_Backend
+{
+
+public:
+
+    // constructor
+    AP_RangeFinder_BLPing(RangeFinder::RangeFinder_State &_state,
+                            AP_RangeFinder_Params &_params,
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: indent issue

Comment (peterbarker) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: ... may also be able to sneak some `const`s in...

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                            AP_SerialManager &serial_manager,
+                            uint8_t serial_instance);
+
+    // static detection function
+    static bool detect(AP_SerialManager &serial_manager, uint8_t serial_instance);
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: This is the pattern, of course, but we now have a singleton for serialmanager.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    // update state
+    void update(void) override;
+
+protected:
+
+    virtual MAV_DISTANCE_SENSOR _get_mav_distance_sensor_type() const override {
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: Sorry for the underscore ;-)

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        return MAV_DISTANCE_SENSOR_ULTRASOUND;
+    }
+
+private:
+
+    // initialise sensor
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: More pointless comments :-)

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    // read a distance from the sensor
+    bool get_reading(uint16_t &reading_cm);
+
+    // process one byte received on serial port
+    // returns true if a distance message has been successfully parsed
+    // state is stored in msg structure
+    bool parse_byte(uint8_t b);
+
+    enum msg_state_t {
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: Make this `enum class` and strip off all of the stuff.

So 
```
enum class msg_state : uint8_t {
   HEADER1,
   HEADER2,
.
.
}
```

and then

```
case msg_state::LEN_L
```

We don't have much precedent for the case of the `enum class`, but I think

```
case MSGState::LEN_L:
```

looks marginally better.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        MSG_STATE_MSG_ID_L,
+        MSG_STATE_MSG_ID_H,
+        MSG_STATE_SRC_ID,
+        MSG_STATE_DST_ID,
+        MSG_STATE_PAYLOAD,
+        MSG_STATE_CRC_L,
+        MSG_STATE_CRC_H
+    };
+
+    AP_HAL::UARTDriver *uart = nullptr;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: Pointless initialisation

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+        break;
+
+    case MSG_STATE_DST_ID:
+        msg.crc_expected += b;
+        msg.state = MSG_STATE_PAYLOAD;
+        break;
+
+    case MSG_STATE_PAYLOAD:
+        if (msg.payload_recv < msg.payload_len) {
+            msg.payload[msg.payload_recv] = b;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: This looks like  a nasty buffer overwrite issue; `msg.payload[]` is only 20 bytes long ATM.  Suggest just throwing away bytes if `msg.payload_recv >= ARRAY_SIZE(msg.payload)`

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    uint32_t last_init_ms;      // system time that sensor was last initialised
+    uint8_t msg_buf[20];        // incoming message buffer
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: Two unused variables here.

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp [hunk 17]
```diff
@@ -6,7 +6,7 @@ const AP_Param::GroupInfo AP_RangeFinder_Params::var_info[] = {
     // @Param: TYPE
     // @DisplayName: Rangefinder type
     // @Description: What type of rangefinder device that is connected
-    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLiteV2-I2C,5:PX4-PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:uLanding,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:BenewakeTFmini
+    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLiteV2-I2C,5:PX4-PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:uLanding,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:BenewakeTFmini,21:LidarLightV3HP,22:PWM,23:BlueRoboticsPing
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp]: Looks like options 21 and 22 should go in regardless of this PR (unrelated change)...

**Comment by jaxxzer** [comment]
Thanks @rmackay9 I'll check on this this week.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_BLPing.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <ctype.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: removed, thanks

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_BLPing.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <ctype.h>
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: moved some includes around, removed unused one

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+// 8-n      uint8_t[]   payload         message payload
+// (n+1)-(n+2)  uint16_t    checksum    the sum of all the non-checksum bytes in the message (low byte, high byte)
+
+/* 
+   The constructor also initialises the rangefinder. Note that this
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the rangefinder
+*/
+AP_RangeFinder_BLPing::AP_RangeFinder_BLPing(RangeFinder::RangeFinder_State &_state,
+                                                             AP_RangeFinder_Params &_params,
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: fixed indentation

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+/*
+   update the state of the sensor
+*/
+void AP_RangeFinder_BLPing::update(void)
+{
+    uint32_t now = AP_HAL::millis();
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: added const

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+// detect if a serial port has been setup to accept rangefinder input
+bool AP_RangeFinder_BLPing::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+/*
+   update the state of the sensor
+*/
+void AP_RangeFinder_BLPing::update(void)
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: txs, added sanity check of uart

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+{
+    uint32_t now = AP_HAL::millis();
+
+    if (get_reading(state.distance_cm)) {
+        // update range_valid state based on distance measured
+        state.last_reading_ms = AP_HAL::millis();
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: fixed, thanks

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+
+        // initialise sensor if no distances recently
+        if (now - last_init_ms > BLPING_INIT_RATE_MS) {
+            last_init_ms = now;
+            init_sensor();
+        }
+    }
+}
+
+// initialise sensor
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: removed pointless comment, txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+void AP_RangeFinder_BLPing::init_sensor()
+{
+    // request distance from sensor
+    send_message(BLDPIN_MSGID_DISTANCE_SIMPLE, nullptr, 0);
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: perhaps although the Lightware serial range finder is similar.  We always send it a "d" if we want a new reading.  I would perhaps also prefer that the sensor just sends data as quickly as it can which result in a faster update rate.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+// send message to sensor
+void AP_RangeFinder_BLPing::send_message(uint16_t msgid, uint8_t *payload, uint16_t payload_len)
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: const added, txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    // write header
+    uart->write((uint8_t)BLPING_FRAME_HEADER1);
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: ok,txs.  check of txspace added

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+        if (b < 0) {
+            continue;
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: txs, replaced continue with break.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+        return false;
+    }
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: const added, txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 2 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: txs, += replaced with |=.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 3 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: replaced, txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 4 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: += replaced, txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 5 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: yes, no recording of parsing errors.. at the risk of passing the buck, I don't think we have reporting of parsing errors from other lidar either..

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 6 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: I tried to make it an enum but it because messy in the parser which could a message id that isn't listed in the enum and I'm not too keen on adding every message id that the sensor supports (there are a lot!)

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 7 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: const added txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 8 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: No I didn't so removed that include, txs

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h (hunk 9 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: fixed indent and added const to params which we never actually use.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h (hunk 10 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: I might leave it accepting the serial manager argument and let someone else blast them for all the drivers at once.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h (hunk 12 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: removed pointless comment

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h (hunk 13 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: as mentioned above I tried to use an enum but ran into troubles in the parser for cases where the message might not be in  the enum

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h (hunk 16 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: ah, nice, removed!

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp (hunk 17 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp]: yes, an unrelated fix I snuck into this PR..

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: https://github.com/ArduPilot/ardupilot/pull/10938/files#diff-808bc25d8005172cc2ec3a3b7de2fac4R33 <- it's only those defines that matter, right?

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h (hunk 14 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.h]: removed pointless initialisation, txs!

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 15 above)
Comment (rmackay9) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: added check before adding new character to the msg structure's payload array.

**Comment by rmackay9** [comment]
Merged!

@jaxxzer, if you (or anyone else including me) finds issues we can resolve them in a follow up PR.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 6 above)
Comment (jaxxzer) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: @rmackay9 You could use an enum and add a `default` handler, no need to handle every message (or every integer in the 16-bit space for that matter).

Comment (jaxxzer) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: You also only need to put the messages you are interested in using in the enum.
